src/sig_stfl/xmss/sig_stfl_xmss_secret_key_functions.c

```diff
@@ -0,0 +1,166 @@
+
+	// Set Secret Key unlocking / releasing function
+	sk->unlock_key = NULL;
+
+	// Set Secret Key saving function
+	sk->secure_store_scrt_key = NULL;
+
+	// Set Secret Key store callback function
+	sk->set_scrt_key_store_cb = OQS_SECRET_KEY_XMSS_set_store_cb;
+
+	// Set Secret Key free function
+	sk->free_key = OQS_SECRET_KEY_XMSS_free;
+
+	return sk;
+}
+
+/* Serialize XMSS secret key data into a byte string, return an allocated buffer. Users have to unallocated the buffer. */
+OQS_STATUS OQS_SECRET_KEY_XMSS_serialize_key(uint8_t **sk_buf_ptr, size_t *sk_len, const OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL || sk_len == NULL || sk_buf_ptr == NULL) {
+		return OQS_ERROR;
+	}
+
+	/* Lock the key if possible */
+	OQS_SECRET_KEY_XMSS_acquire_lock(sk);
+
+	uint8_t *sk_buf = malloc(sk->length_secret_key * sizeof(uint8_t));
+	if (sk_buf == NULL) {
+		return OQS_ERROR;
+	}
+
+	// Simply copy byte string of secret_key_data
+	memcpy(sk_buf, sk->secret_key_data, sk->length_secret_key);
+
+	*sk_buf_ptr = sk_buf;
+	*sk_len = sk->length_secret_key;
+
+	/* Unlock the key if possible */
+	OQS_SECRET_KEY_XMSS_release_lock(sk);
+
+	return OQS_SUCCESS;
+}
+
+/* Only for internal use. Similar to OQS_SECRET_KEY_XMSS_serialize_key, but this function does not aquire and release lock. */
+OQS_STATUS OQS_SECRET_KEY_XMSS_inner_serialize_key(uint8_t **sk_buf_ptr, size_t *sk_len, const OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL || sk_len == NULL || sk_buf_ptr == NULL) {
+		return OQS_ERROR;
+	}
+
+	uint8_t *sk_buf = malloc(sk->length_secret_key * sizeof(uint8_t));
+	if (sk_buf == NULL) {
+		return OQS_ERROR;
+	}
+
+	// Simply copy byte string of secret_key_data
+	memcpy(sk_buf, sk->secret_key_data, sk->length_secret_key);
+
+	*sk_buf_ptr = sk_buf;
+	*sk_len = sk->length_secret_key;
+
+	return OQS_SUCCESS;
+}
+
+/* Deserialize XMSS byte string into an XMSS secret key data. */
+OQS_STATUS OQS_SECRET_KEY_XMSS_deserialize_key(OQS_SIG_STFL_SECRET_KEY *sk, const size_t sk_len, const uint8_t *sk_buf, XMSS_UNUSED_ATT void *context) {
+#ifndef OQS_ALLOW_XMSS_KEY_AND_SIG_GEN
+	return OQS_ERROR;
+#endif
+
+	if (sk == NULL || sk_buf == NULL || (sk_len != sk->length_secret_key)) {
+		return OQS_ERROR;
+	}
+
+	memcpy(sk->secret_key_data, sk_buf, sk->length_secret_key);
+	sk->context = context;
+
+	return OQS_SUCCESS;
+}
+
+void OQS_SECRET_KEY_XMSS_set_store_cb(OQS_SIG_STFL_SECRET_KEY *sk, secure_store_sk store_cb, void *context) {
+	if (sk == NULL || store_cb == NULL) {
+		return;
+	}
+	sk->secure_store_scrt_key = store_cb;
+	sk->context = context;
+}
+
+void OQS_SECRET_KEY_XMSS_free(OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL) {
+		return;
+	}
+
+	OQS_MEM_secure_free(sk->secret_key_data, sk->length_secret_key);
+	sk->secret_key_data = NULL;
+}
+
+void OQS_SECRET_KEY_XMSS_acquire_lock(const OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL) {
+		return;
+	}
+
+	/* Lock the key if possible */
+	if ((sk->lock_key != NULL) && (sk->mutex != NULL)) {
+		sk->lock_key(sk->mutex);
```

Comment:
Why isn't this result being error checked? And why is `_acquire_lock` also not returning a status?

Reply:
Thanks @dstebila , that's right, the lock must be checked. 

I added the check in #1712 .

Reply:
Fixed in #1712